src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/SpatioTemporalZonalStatistics.java

```diff
@@ -0,0 +1,583 @@
+                GridCoverage2DReader reader,
+                SimpleFeatureCollection zones,
+                List<Object> times,
+                Set<Statistic> requestedStats) {
+            super(zones);
+            this.reader = reader;
+            this.times = times;
+            this.requestedStats = requestedStats;
+
+            SimpleFeatureTypeBuilder tb = new SimpleFeatureTypeBuilder();
+            for (AttributeDescriptor att : zones.getSchema().getAttributeDescriptors()) {
+                tb.minOccurs(att.getMinOccurs());
+                tb.maxOccurs(att.getMaxOccurs());
+                tb.restrictions(att.getType().getRestrictions());
+                if (att instanceof GeometryDescriptor) {
+                    GeometryDescriptor gatt = (GeometryDescriptor) att;
+                    tb.crs(gatt.getCoordinateReferenceSystem());
+                }
+                tb.add("z_" + att.getLocalName(), att.getType().getBinding());
+            }
+
+            addAttributes(tb, requestedStats);
+            tb.setName(zones.getSchema().getName());
+            targetSchema = tb.buildFeatureType();
+        }
+
+        @Override
+        public SimpleFeatureType getSchema() {
+            return targetSchema;
+        }
+
+        @Override
+        public SimpleFeatureIterator features() {
+            return new SpatioTemporalZonalStatisticsIterator(
+                    delegate.features(), reader, targetSchema, times, requestedStats);
+        }
+    }
+
+    /**
+     * An iterator computing statistics as we go
+     *
+     * <p>Part of this code has been imported and adapted from RasterZonalStatistics
+     */
+    static class SpatioTemporalZonalStatisticsIterator implements SimpleFeatureIterator {
+
+        private final List<Object> times;
+        private GridCoverage2DReader reader;
+        private SimpleFeatureIterator zones;
+        private SimpleFeatureBuilder builder;
+        private List<SimpleFeature> features = new ArrayList<>();
+        private Set<Statistic> requestedStats;
+
+        public SpatioTemporalZonalStatisticsIterator(
+                SimpleFeatureIterator zones,
+                GridCoverage2DReader reader,
+                SimpleFeatureType targetSchema,
+                List<Object> times,
+                Set<Statistic> requestedStats) {
+            this.zones = zones;
+            this.builder = new SimpleFeatureBuilder(targetSchema);
+            this.reader = reader;
+            this.requestedStats = requestedStats;
+            this.times = times;
+        }
+
+        @Override
+        public void close() {
+            zones.close();
+        }
+
+        @Override
+        public boolean hasNext() {
+            return !features.isEmpty() || zones.hasNext();
+        }
+
+        @Override
+        public SimpleFeature next() throws NoSuchElementException {
+            // build the next set of features if necessary
+            if (features.isEmpty()) {
```

Comment:
How do ever get more than one feature in this buffer? It would be simpler to understand if it was a ``SimpleFeature nextFeature``, null or not.

Reply:
Fixed